### PR TITLE
`balanceTx` uses pure `TimeTranslation` instead of an impure `TimeInterpreter`

### PR DESCRIPTION
**Pull request by Unisay**
- [x] Create `TimeTranslation` opaque data type as a purer replacement for the `TimeInterpreter`
- [x] Provide a [temporary] constructor to create `TimeTranslation` from a `TimeInterpreter`
- [x] Replace usages of `TimeInterpreter` with the `TimeTranslation`

### Issue Number

ADP-2968

### CODE REVIEW

File: lib/wallet/src/Cardano/Wallet.hs [hunk 1]
```diff
@@ -2087,10 +2087,8 @@ buildSignSubmitTransaction ti db@DBLayer{..} netLayer txLayer pwd walletId
             & throwWrappedErr wrapNetworkError
             & liftIO
 
-        slotToUTCTime slot
-            & interpretQuery (neverFails "slot is ahead of the node tip" ti)
-            & fmap (BuiltTx{..},)
-            & liftIO
+        utcTime <- translateSlotToUTC timeTranslation slot & either throwIO pure
```
Comment (Anviking) [inline lib/wallet/src/Cardano/Wallet.hs]: Could we keep using `TimeInterpreter` here to be consistent with the rest of the wallet?

File: lib/wallet/src/Cardano/Wallet/Primitive/Types/TimeTranslation.hs [hunk 2]
```diff
@@ -0,0 +1,58 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StrictData #-}
+
+module Cardano.Wallet.Primitive.Types.TimeTranslation
```
Comment (Anviking) [inline lib/wallet/src/Cardano/Wallet/Primitive/Types/TimeTranslation.hs]:
```suggestion
module Cardano.Wallet.Write.Tx.TimeTranslation
```

**Comment by Anviking** [comment]
>  Deprecate and remove TimeInterpreter abstraction.

🤔 I don't think that's a good idea as part of this task. The wallet's needs are different from that of `balanceTx`. For example `TimeInterpreter` currently handles logging in a consistent manner.

**Comment by Unisay** [comment]
> > Deprecate and remove TimeInterpreter abstraction.
> 
> thinking I don't think that's a good idea as part of this task. The wallet's needs are different from that of `balanceTx`. For example `TimeInterpreter` currently handles logging in a consistent manner.

Agreed.

File: lib/wallet/src/Cardano/Wallet.hs (hunk 1 above)
Comment (Unisay) [reply lib/wallet/src/Cardano/Wallet.hs]: Rolled back.

File: lib/wallet/src/Cardano/Wallet/Primitive/Slotting.hs [hunk 3]
```diff
@@ -401,6 +406,14 @@ getSystemStart :: TimeInterpreter m -> SystemStart
 getSystemStart TimeInterpreter{blockchainStartTime} =
     let (StartTime t) = blockchainStartTime in SystemStart t
 
+toTimeTranslation
+    :: TimeInterpreter (ExceptT PastHorizonException IO)
+    -> IO TimeTranslation
+toTimeTranslation timeInterpreter = do
```
Comment (Anviking) [inline lib/wallet/src/Cardano/Wallet/Primitive/Slotting.hs:412]: Ah, you moved it here to invert the dependency —  great!

File: lib/wallet/src/Cardano/Wallet/Primitive/Slotting.hs [hunk 4]
```diff
@@ -401,6 +406,14 @@ getSystemStart :: TimeInterpreter m -> SystemStart
...
+    info <- runExceptT (toEpochInfo timeInterpreter)
+        >>= either throwIO (pure . hoistEpochInfo runExcept)
```
Comment (Anviking) [inline lib/wallet/src/Cardano/Wallet/Primitive/Slotting.hs:414]: In practice, we should never hit this `throwIO` per the removed comment from `Shelley.Transaction` Possibly we could still hint at this being the case.

**Review by Anviking — APPROVED** [review]

**Comment by Unisay** [comment]
bors merge

**Comment by Unisay** [comment]
bors merge

**Pull request merged**